### PR TITLE
Auto remove tracking elements from shared URLs

### PR DESCRIPTION
**Pull request by pawisoon**
### First time contributor checklist

- [x] I have read the [README](https://github.com/signalapp/Signal-iOS/blob/main/README.md) and [CONTRIBUTING](https://github.com/signalapp/Signal-iOS/blob/main/CONTRIBUTING.md) documents
- [x] I have signed the [Contributor Licence Agreement](https://signal.org/cla/)

### Contributor checklist

- [x] My commits are rebased on the latest main branch
- [x] My commits are in nice logical chunks
- [x] My contribution is fully baked and is ready to be merged as is
- [x] I have tested my contribution on these devices:
 * iPhone 15, iOS 17.0
 * iPhone 13, iOS 15.5

- - - - - - - - - -

### Description

As a privacy-oriented user, I would like to have the option to remove tracking elements from URLs pasted / shared by me into my chats. 

This PR introduces a feature that removes tracking elements from URL that are pasted or shared using ShareExtension.
Removed elements and platforms: 
- \"igshid=\" for Instagram
- \"si=\" for Youtube
- \"fbclid=\" for Facebook
- \"t=\" removed only for Twitter(X)
-  \"ttclid=\" for TikTok Click ID
- \"gclid=\" for Google
- other tracking elements \"utm_campaign\", \"utm_medium\" and \"utm_source\"

### Example 
Example of shared URL that has a source identifier "?si=": 

`https://youtu.be/dQw4w9WgXcQ?si=ZvcTJw_HfIbBdVek`

Not tracked URL should look like this:

`https://youtu.be/dQw4w9WgXcQ`

 Feature is optional for user (disabled by default) and can be enabled in App Settings as explained below: 
`App Seettings -> Privacy -> Advanced -> "Remove Tracking Elements"`

settings

## Testing: 
The feature was tested manually by sharing a URL that has a tracking element in it. 
It was tested in the following 2 scenarios.

### Scenario A: 
URL stored in clipboard of iOS device is pasted in chat. If setting is enabled and pasted URL contains tracking elements, then its automatically removing trackers and changing URL in text view. 

scenario_a

### Scenario B: 
User attempts to share a URL from within an app e.g. youtube using ShareExtension. 
If setting is enabled, and shared URL contains tracking elements, then its automatically removing trackers and changing URL in share preview text view. 

scenario_b

### CODE REVIEW

**Comment by dhanishgajjar** [comment]
Would love to have this feature 👌🏽

**Comment by michtesar** [comment]
I like this! Very application should handle links with privacy in mind! Thanks 👍

**Comment by pawisoon** [comment]
@Miguel-Signal @harry-signal @amy-signal can this be reviewed? :)

**Comment by RemainingToast** [comment]
This looks good +1

But it should remove the "?" as a qol

**Comment by pawisoon** [comment]
> This looks good +1
> 
> But it should remove the "?" as a qol

Thanks for feedback! I pushed change that fixes it 😄

**Comment by DGrothe-PhD** [comment]
Perhaps `t=` shouldn't be removed for youtube links, consider https://youtu.be/x0EYpi38Yp4?t=931 for example. Therein `t`  stands for time to start from.

**Comment by pawisoon** [comment]
> Perhaps `t=` shouldn't be removed for youtube links, consider https://youtu.be/x0EYpi38Yp4?t=931 for example. Therein `t` stands for time to start from.

Thank you!  It's now fixed to remove "t=" parameter only for Twitter/X hostnames.

File: Signal/src/ViewControllers/ConversationView/ConversationInputTextView.swift [hunk 1]
```diff
@@ -255,12 +255,24 @@ extension UITextView {
...
+    
     var components = URLComponents(url: url, resolvingAgainstBaseURL: true)
     let items = components?.queryItems
     components?.queryItems = items?.filter { item in
         !paramsToRemove.contains(item.name)
     }
+    
+    // Remove "t=" parameter only for twitter/x
+    if let host = components?.host, twitterHosts.contains(where: { $0 == host }) {
+        components?.queryItems = items?.filter { item in
```
[the remark this answers is not in the document]
Comment (pawisoon) [reply Signal/src/ViewControllers/ConversationView/ConversationInputTextView.swift:270]: With your suggestion applied I receive the following error: 
`Overlapping accesses to 'components', but modification requires exclusive access; consider copying to a local variable`

That's why it's stored in `let items`

**Comment by pawisoon** [comment]
I'm sorry for pinging multiple people but I don't know who can have a look at my PR? 
@igor-signal @pete-signal @max-signal @marissa-signal

**Comment by schmeat** [comment]
@pawisoon For what it's worth, I think this is a great feature and you've done a good job on it. However, I wouldn't expect any interaction from the signal team. There is a graveyard of pull requests with no interaction on this project.

**Comment by harry-signal** [comment]
We greatly appreciate your effort and that of past contributions on this subject! Please understand, though, that for some sensitive subjects we are unlikely to accept external contributions as-is. That doesn't meant we are ignoring them; your contributions are meaningful and lead to internal discussions that drive our future roadmap, especially for tricky areas like this one.

**Comment by pawisoon** [comment]
@harry-signal Thank for your reply Harry. Appreciate it :) 
I really hope this will get merged into Signal app and will make it even more privacy-oriented app for its users.

**Comment by elcodedocle** [comment]
+1 on this really cool feature!

**Pull request closed without merging**